File: lambda/node/services/jobs.js
const { defaultHeaders } = require("../lib/response.js");
const crypto = require("crypto");

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { PutCommand, GetCommand, UpdateCommand, DeleteCommand, QueryCommand, DynamoDBDocumentClient } = require("@aws-sdk/lib-dynamodb");

const client = new DynamoDBClient({ region: "us-east-1" });
const ddb = DynamoDBDocumentClient.from(client);

const jobTableName = process.env.JOBS_TABLE_NAME;

exports.getJobs = async (event) => {
  const userId = event.requestContext.authorizer.claims.sub;

  const getJobsCommand = new QueryCommand({
    TableName: jobTableName,
    KeyConditionExpression: "userId = :uid",
    ExpressionAttributeValues: {
      ":uid": userId,
    },
  });

  try {
    const result = await ddb.send(getJobsCommand);
    const allJobs = result.Items || [];
    const jobsNoUid = allJobs.map(({userId, ...rest}) => rest);

    return {
      statusCode: 200,
      headers: defaultHeaders,
      body: JSON.stringify(jobsNoUid),
    };
  } catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: defaultHeaders,
      body: JSON.stringify({ error: "Failed to get job applications" }),
    };
  }

}

exports.getJobById = async (event, id) => {
  const userId = event.requestContext.authorizer.claims.sub;

  const getJobByIdCommand = new GetCommand({
    TableName: jobTableName,
    Key: {
      userId: userId,
      jobId: id
    }
  });

  try {
    const result = await ddb.send(getJobByIdCommand);

    return {
      statusCode: 200,
      headers: defaultHeaders,
      body: JSON.stringify(result.Items),
    };
  } catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: defaultHeaders,
      body: JSON.stringify({ error: "Failed to get job application" }),
    };
  }
};

exports.createJob = async (event) => {
  const userId = event.requestContext.authorizer.claims.sub;
  const newApp = JSON.parse(event.body || '{}');
  const jobId = crypto.randomUUID();

  const newJob = {
    userId,
    jobId,
    company: newApp.company,
    position: newApp.position,
    status: newApp.status,
    dateApplied: newApp.dateApplied || new Date().toISOString().split("T")[0],
  };

  try {
    await ddb.send(new PutCommand({
      TableName: jobTableName,
      Item: newJob,
    }));

    const { userId, ...newJobNoUid } = newJob;

    return {
      statusCode: 201,
      headers: defaultHeaders,
      body: JSON.stringify(newJobNoUid),
    };
  } catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: defaultHeaders,
      body: JSON.stringify({ error: "Failed to save job application" }),
    };
  }
};

exports.updateJob = async (event, id) => {
  const userId = event.requestContext.authorizer.claims.sub;
  const updates = JSON.parse(event.body || '{}');

  const updateExpressions = [];
  const expressionValues  = {};
  const expressionNames = {};
  for (const key in updates) {
    updateExpressions.push(`#${key} = :${key}`);
    expressionValues[`:${key}`] = updates[key];
    expressionNames[`#${key}`] = key
  }

  const command = new UpdateCommand({
    TableName: jobTableName,
    Key: {
      userId: userId,
      jobId: id
    },
    UpdateExpression: `SET ${updateExpressions.join(", ")}`,
    ExpressionAttributeValues: expressionValues,
    ExpressionAttributeNames: expressionNames,
    ReturnValues: "ALL_NEW",
  });

  try {
    await ddb.send(command);

    return {
      statusCode: 200,
      headers: defaultHeaders,
      body: JSON.stringify({ message: "Update successful"})
    }
  } catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: defaultHeaders,
      body: JSON.stringify({ error: "Failed to update job application"})
    }
  }
};

exports.deleteJob = async (event, id) => {
  const userId = event.requestContext.authorizer.claims.sub;
  
  const deleteJobByIdCommand = new DeleteCommand({
    TableName: jobTableName,
    Key: {
      userId: userId,
      jobId: id
    }
  });

  try {
    const result = await ddb.send(deleteJobByIdCommand);

    return {
      statusCode: 204,
      headers: defaultHeaders,
      body: JSON.stringify({ message: `Job ${id} successfully deleted` }),
    };
  } catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: defaultHeaders,
      body: JSON.stringify({ error: "Failed to delete job application" }),
    };
  }
};

exports.updateJobStatus = async (event, id) => {
  const userId = event.requestContext.authorizer.claims.sub;
  const { status } = JSON.parse(event.body || {});

  const allowedStatuses = ['Applied', 'Interview', 'Offer', 'Rejected'];
  if (!allowedStatuses.includes(status)) {
    return {
      statusCode: 400,
      body: JSON.stringify({ error: 'Invalid status value' }),
    };
  }

  const command = new UpdateCommand({
    TableName:jobTableName,
    Key: {
      userId: userId,
      jobId: jobId,
    },
    UpdateExpression: "SET #s = :newStatus",
    ExpressionAttributeNames: {
      "#s": "status",
    },
    ExpressionAttributeValues: {
      ":newStatus": newStatus,
    },
    ReturnValues: "ALL_NEW"
  });

  try {
    const result = await ddb.send(command);

    return {
      statusCode: 200,
      headers: defaultHeaders,
      body: JSON.stringify(result.Attr)
    }
  } catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: defaultHeaders,
      body: JSON.stringify({ error: "Failed to update job application"})
    }
  }

}

File: lambda/python/jobs/services/jobs.py
import os
import boto3
import json
import uuid
from datetime import datetime
from boto3.dynamodb.conditions import Key
from ..lib.response import default_headers

dynamodb = boto3.resource("dynamodb", region_name="us-east-1")

def get_table():
    return dynamodb.Table(os.environ["JOBS_TABLE_NAME"])

def get_jobs(event):
    user_id = event["requestContext"]["authorizer"]["claims"]["sub"]

    try:
        table = get_table()
        response = table.query(KeyConditionExpression=Key("userId").eq(user_id))
        items = response.get("Items", [])
        jobs_no_uid = [{k: v for k, v in item.items() if k != "userId"} for item in items]

        return {"statusCode": 200, "headers": default_headers, "body": json.dumps(jobs_no_uid)}
    except Exception as e:
        print(e)
        return {"statusCode": 500, "headers": default_headers, "body": json.dumps({"error": "Failed to get job applications"})}


def get_job_by_id(event, job_id):
    user_id = event["requestContext"]["authorizer"]["claims"]["sub"]

    try:
        table = get_table()
        result = table.get_item(Key={"userId": user_id, "jobId": job_id})
        return {"statusCode": 200, "headers": default_headers, "body": json.dumps(result.get("Item"))}
    except Exception as e:
        print(e)
        return {"statusCode": 500, "headers": default_headers, "body": json.dumps({"error": "Failed to get job application"})}


def create_job(event):
    user_id = event["requestContext"]["authorizer"]["claims"]["sub"]
    body = json.loads(event.get("body", "{}"))

    job_id = str(uuid.uuid4())
    new_job = {
        "userId": user_id,
        "jobId": job_id,
        "company": body.get("company"),
        "position": body.get("position"),
        "status": body.get("status"),
        "dateApplied": body.get("dateApplied") or datetime.utcnow().strftime("%Y-%m-%d"),
    }

    try:
        table = get_table()
        table.put_item(Item=new_job)
        job_no_uid = {k: v for k, v in new_job.items() if k != "userId"}

        return {"statusCode": 201, "headers": default_headers, "body": json.dumps(job_no_uid)}
    except Exception as e:
        print(e)
        return {"statusCode": 500, "headers": default_headers, "body": json.dumps({"error": "Failed to save job application"})}


def update_job(event, job_id):
    user_id = event["requestContext"]["authorizer"]["claims"]["sub"]
    updates = json.loads(event.get("body", "{}"))

    update_expressions = []
    expression_values = {}
    expression_names = {}

    for key, value in updates.items():
        update_expressions.append(f"#{key} = :{key}")
        expression_values[f":{key}"] = value
        expression_names[f"#{key}"] = key

    try:
        table = get_table()
        table.update_item(
            Key={"userId": user_id, "jobId": job_id},
            UpdateExpression="SET " + ", ".join(update_expressions),
            ExpressionAttributeValues=expression_values,
            ExpressionAttributeNames=expression_names,
            ReturnValues="ALL_NEW",
        )

        return {"statusCode": 200, "headers": default_headers, "body": json.dumps({"message": "Update successful"})}
    except Exception as e:
        print(e)
        return {"statusCode": 500, "headers": default_headers, "body": json.dumps({"error": "Failed to update job application"})}


def delete_job(event, job_id):
    user_id = event["requestContext"]["authorizer"]["claims"]["sub"]

    try:
        table = get_table()
        table.delete_item(Key={"userId": user_id, "jobId": job_id})

        return {"statusCode": 204, "headers": default_headers, "body": json.dumps({"message": f"Job {job_id} successfully deleted"})}
    except Exception as e:
        print(e)
        return {"statusCode": 500, "headers": default_headers, "body": json.dumps({"error": "Failed to delete job application"})}


def update_job_status(event, job_id):
    user_id = event["requestContext"]["authorizer"]["claims"]["sub"]
    status = json.loads(event.get("body", "{}"))['status']

    allowed_statuses = ["Applied", "Interview", "Offer", "Rejected"]
    if status not in allowed_statuses:
        return {"statusCode": 400, "headers": default_headers, "body": json.dumps({"error": "Invalid status value"})}

    try:
        table = get_table()
        result = table.update_item(
            Key={"userId": user_id, "jobId": job_id},
            UpdateExpression="SET #s = :newStatus",
            ExpressionAttributeNames={"#s": "status"},
            ExpressionAttributeValues={":newStatus": status},
            ReturnValues="ALL_NEW",
        )

        return {"statusCode": 200, "headers": default_headers, "body": json.dumps(result.get("Attributes"))}
    except Exception as e:
        print(e)
        return {"statusCode": 500, "headers": default_headers, "body": json.dumps({"error": "Failed to update job application"})}
